{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-18T08:46:39.109669Z",
     "start_time": "2019-09-18T08:46:39.075118Z"
    }
   },
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'script_utils'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-15-463a101414d2>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0msys\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     19\u001b[0m \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'../code'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 20\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mscript_utils\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mshow_output\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     21\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mcode\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mp3\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mrun_primer3\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     22\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'script_utils'"
     ]
    }
   ],
   "source": [
    "# home\n",
    "home = '/Users/mahtin'\n",
    "home = '/Users/martinscience'\n",
    "\n",
    "\n",
    "import xlrd\n",
    "import os\n",
    "\n",
    "from IPython.display import display\n",
    "pd.set_option('display.max_columns', None)\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "InteractiveShell.ast_node_interactivity = \"all\"\n",
    "pd.set_option('max_colwidth', 200)\n",
    "static_path = os.path.join(home, \"Dropbox/Icke/Work/static\")\n",
    "\n",
    "\n",
    "# get the code\n",
    "import sys\n",
    "sys.path.append('../code')\n",
    "from script_utils import show_output\n",
    "from code.p3 import run_primer3\n",
    "\n",
    "\n",
    "static_path = os.path.join(home, \"Dropbox/Icke/Work/static\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load mutation list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CoopPrimerList.xlsx\r\n"
     ]
    }
   ],
   "source": [
    "ls /Users/martinscience/Desktop/NEBnextCoop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UPN</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Exon</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Mut Start</th>\n",
       "      <th>Mut End</th>\n",
       "      <th>Start Forward-Primer</th>\n",
       "      <th>End Forward-Primer</th>\n",
       "      <th>Forward Primer 5'-3'</th>\n",
       "      <th>Start Reverse-Primer</th>\n",
       "      <th>End Reverse-Primer</th>\n",
       "      <th>Reverse Primer 5'-3'</th>\n",
       "      <th>Product size (bp)</th>\n",
       "      <th>Comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>DNAH5</td>\n",
       "      <td>NM_001369</td>\n",
       "      <td>exon28</td>\n",
       "      <td>chr5</td>\n",
       "      <td>13864642</td>\n",
       "      <td>13864642</td>\n",
       "      <td>13864704</td>\n",
       "      <td>13864723</td>\n",
       "      <td>TGAAGGACTGGCAGGCTTTT</td>\n",
       "      <td>13864569</td>\n",
       "      <td>13864588</td>\n",
       "      <td>ATTCCCCACATCCAGACTGT</td>\n",
       "      <td>155</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>DNAH5</td>\n",
       "      <td>NM_001369</td>\n",
       "      <td>exon35</td>\n",
       "      <td>chr5</td>\n",
       "      <td>13839514</td>\n",
       "      <td>13839514</td>\n",
       "      <td>13839596</td>\n",
       "      <td>13839615</td>\n",
       "      <td>CCATGGACTTTGAGTGGCTG</td>\n",
       "      <td>13839441</td>\n",
       "      <td>13839460</td>\n",
       "      <td>GGTTGGGGAGAAGGGTTCC</td>\n",
       "      <td>175</td>\n",
       "      <td>R-Primer intronic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>DNAH5</td>\n",
       "      <td>NM_001369</td>\n",
       "      <td>exon71</td>\n",
       "      <td>chr5</td>\n",
       "      <td>13721298</td>\n",
       "      <td>13721298</td>\n",
       "      <td>13721335</td>\n",
       "      <td>13721352</td>\n",
       "      <td>GCCCGCAAGTACATCGTG</td>\n",
       "      <td>13721196</td>\n",
       "      <td>13721215</td>\n",
       "      <td>CTTCCCCAAGGCAATGATGG</td>\n",
       "      <td>159</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>14</td>\n",
       "      <td>DNAH5</td>\n",
       "      <td>NM_001369</td>\n",
       "      <td>exon76</td>\n",
       "      <td>chr5</td>\n",
       "      <td>13708242</td>\n",
       "      <td>13708242</td>\n",
       "      <td>13708323</td>\n",
       "      <td>13708342</td>\n",
       "      <td>ACCCTCACTGAGCTGAAACT</td>\n",
       "      <td>13708154</td>\n",
       "      <td>13708174</td>\n",
       "      <td>TCATGCTTTCCACTTGCCAA</td>\n",
       "      <td>188</td>\n",
       "      <td>R-Primer intronic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>WSCD2</td>\n",
       "      <td>NM_014653</td>\n",
       "      <td>exon2</td>\n",
       "      <td>chr12</td>\n",
       "      <td>108589770</td>\n",
       "      <td>108589770</td>\n",
       "      <td>108589697</td>\n",
       "      <td>108589716</td>\n",
       "      <td>GGGAGCCTTGTCTTCCTTCA</td>\n",
       "      <td>108589834</td>\n",
       "      <td>108589853</td>\n",
       "      <td>CTTCACCTGTGTCCCGGAAA</td>\n",
       "      <td>157</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>WSCD2</td>\n",
       "      <td>NM_014653</td>\n",
       "      <td>exon3</td>\n",
       "      <td>chr12</td>\n",
       "      <td>108600174</td>\n",
       "      <td>108600174</td>\n",
       "      <td>108600069</td>\n",
       "      <td>108600088</td>\n",
       "      <td>AGTACATCGGCTGCTACCTG</td>\n",
       "      <td>108600194</td>\n",
       "      <td>108600210</td>\n",
       "      <td>TGGGAATCTGAGAAGGAAGGG</td>\n",
       "      <td>169</td>\n",
       "      <td>R-Primer intronic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>WSCD2</td>\n",
       "      <td>NM_014653</td>\n",
       "      <td>exon5</td>\n",
       "      <td>chr12</td>\n",
       "      <td>108618575</td>\n",
       "      <td>108618575</td>\n",
       "      <td>108618457</td>\n",
       "      <td>108618476</td>\n",
       "      <td>TCCCAACCAGATCTGACTCC</td>\n",
       "      <td>108618605</td>\n",
       "      <td>108618624</td>\n",
       "      <td>AAGTCCACGCATTTGTCCAC</td>\n",
       "      <td>168</td>\n",
       "      <td>F-Primer intronic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>RB1</td>\n",
       "      <td>NM_000321</td>\n",
       "      <td>exon6</td>\n",
       "      <td>chr13</td>\n",
       "      <td>48923147</td>\n",
       "      <td>48923147</td>\n",
       "      <td>48923119</td>\n",
       "      <td>48923141</td>\n",
       "      <td>GGTGCTAAAAGTTTCTTGGATCA</td>\n",
       "      <td>48923242</td>\n",
       "      <td>48923259</td>\n",
       "      <td>GCAGAGAATGAGGGAGGAGT</td>\n",
       "      <td>158</td>\n",
       "      <td>F-Primer close to Mutation; R-Primer intronic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>ANKMY2</td>\n",
       "      <td>NM_020319</td>\n",
       "      <td>exon3</td>\n",
       "      <td>chr7</td>\n",
       "      <td>16666689</td>\n",
       "      <td>16666689</td>\n",
       "      <td>16666758</td>\n",
       "      <td>16666780</td>\n",
       "      <td>TGCAGCATATAAAGGAAAACTCG</td>\n",
       "      <td>16666663</td>\n",
       "      <td>16666682</td>\n",
       "      <td>ACCAGAAAGTGCAGCAAACA</td>\n",
       "      <td>118</td>\n",
       "      <td>R-Primer very close to mutation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>PCDHA5</td>\n",
       "      <td>NM_018908</td>\n",
       "      <td>exon1</td>\n",
       "      <td>chr5</td>\n",
       "      <td>140203618</td>\n",
       "      <td>140203618</td>\n",
       "      <td>140203570</td>\n",
       "      <td>140203553</td>\n",
       "      <td>GGCAAGCCCACTCTGTTG</td>\n",
       "      <td>140203686</td>\n",
       "      <td>140203712</td>\n",
       "      <td>CGTGGGTTGTCTGTAGAGGT</td>\n",
       "      <td>158</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>RAD54B</td>\n",
       "      <td>NM_001205263</td>\n",
       "      <td>exon4</td>\n",
       "      <td>chr8</td>\n",
       "      <td>95416375</td>\n",
       "      <td>95416375</td>\n",
       "      <td>95416435</td>\n",
       "      <td>95416455</td>\n",
       "      <td>TGCCACGACCAGATAAGAATC</td>\n",
       "      <td>95416312</td>\n",
       "      <td>95416329</td>\n",
       "      <td>TCCTACCTCATTCCCATTACACA</td>\n",
       "      <td>157</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>DDB1</td>\n",
       "      <td>NM_001923</td>\n",
       "      <td>exon6</td>\n",
       "      <td>chr11</td>\n",
       "      <td>61093159</td>\n",
       "      <td>61093159</td>\n",
       "      <td>61093221</td>\n",
       "      <td>61093240</td>\n",
       "      <td>CAGTCCTGGTTTGGCAGAC</td>\n",
       "      <td>61093080</td>\n",
       "      <td>61093099</td>\n",
       "      <td>ACTGCGCTCACCTTGATGAT</td>\n",
       "      <td>169</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "      <td>TP53</td>\n",
       "      <td>NM_000546</td>\n",
       "      <td>exon5</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7578457</td>\n",
       "      <td>7578457</td>\n",
       "      <td>7578530</td>\n",
       "      <td>7578549</td>\n",
       "      <td>CCCTGCCCTCAACAAGATGT</td>\n",
       "      <td>7578364</td>\n",
       "      <td>7578383</td>\n",
       "      <td>TGCTCACCATCGCTATCTGA</td>\n",
       "      <td>186</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13</td>\n",
       "      <td>8</td>\n",
       "      <td>TP53</td>\n",
       "      <td>NM_000546</td>\n",
       "      <td>exon6</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7578268</td>\n",
       "      <td>7578268</td>\n",
       "      <td>7578300</td>\n",
       "      <td>7578319</td>\n",
       "      <td>GGCCTCTGATTCCTCACTGA</td>\n",
       "      <td>7578164</td>\n",
       "      <td>7578183</td>\n",
       "      <td>TTGCAAACCAGACCTCAGGC</td>\n",
       "      <td>156</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>TP53</td>\n",
       "      <td>NM_000546</td>\n",
       "      <td>exon7</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7577538</td>\n",
       "      <td>7577538</td>\n",
       "      <td>7577591</td>\n",
       "      <td>7577610</td>\n",
       "      <td>AGGTTGGCTCTGACTGTACC</td>\n",
       "      <td>7577431</td>\n",
       "      <td>7577452</td>\n",
       "      <td>GGGTCAGAGGCAAGCAGA</td>\n",
       "      <td>178</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15</td>\n",
       "      <td>8</td>\n",
       "      <td>TP53</td>\n",
       "      <td>NM_000546</td>\n",
       "      <td>exon7</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7577570</td>\n",
       "      <td>7577570</td>\n",
       "      <td>7577591</td>\n",
       "      <td>7577610</td>\n",
       "      <td>AGGTTGGCTCTGACTGTACC</td>\n",
       "      <td>7577431</td>\n",
       "      <td>7577452</td>\n",
       "      <td>GGGTCAGAGGCAAGCAGA</td>\n",
       "      <td>178</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      <td>9</td>\n",
       "      <td>TP53</td>\n",
       "      <td>NM_000546</td>\n",
       "      <td>exon5</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7578406</td>\n",
       "      <td>7578406</td>\n",
       "      <td>7578530</td>\n",
       "      <td>7578549</td>\n",
       "      <td>CCCTGCCCTCAACAAGATGT</td>\n",
       "      <td>7578364</td>\n",
       "      <td>7578383</td>\n",
       "      <td>TGCTCACCATCGCTATCTGA</td>\n",
       "      <td>186</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      <td>14</td>\n",
       "      <td>TP53</td>\n",
       "      <td>NM_000546</td>\n",
       "      <td>exon6</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7578253</td>\n",
       "      <td>7578253</td>\n",
       "      <td>7578300</td>\n",
       "      <td>7578319</td>\n",
       "      <td>GGCCTCTGATTCCTCACTGA</td>\n",
       "      <td>7578164</td>\n",
       "      <td>7578183</td>\n",
       "      <td>TTGCAAACCAGACCTCAGGC</td>\n",
       "      <td>156</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      <td>14</td>\n",
       "      <td>TP53</td>\n",
       "      <td>NM_000546</td>\n",
       "      <td>exon5</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7578442</td>\n",
       "      <td>7578442</td>\n",
       "      <td>7578530</td>\n",
       "      <td>7578549</td>\n",
       "      <td>CCCTGCCCTCAACAAGATGT</td>\n",
       "      <td>7578364</td>\n",
       "      <td>7578383</td>\n",
       "      <td>TGCTCACCATCGCTATCTGA</td>\n",
       "      <td>186</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19</td>\n",
       "      <td>14</td>\n",
       "      <td>TP53</td>\n",
       "      <td>NM_000546</td>\n",
       "      <td>exon8</td>\n",
       "      <td>chr17</td>\n",
       "      <td>8000000</td>\n",
       "      <td>7577121</td>\n",
       "      <td>7577185</td>\n",
       "      <td>7577205</td>\n",
       "      <td>GGACCTGATTTCCTTACTGCC</td>\n",
       "      <td>7576997</td>\n",
       "      <td>7577018</td>\n",
       "      <td>GCTTCTTGTCCTGCTTGCTT</td>\n",
       "      <td>209</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    UPN    Gene    Transcript    Exon    Chr  Mut Start    Mut End   \\\n",
       "0     3   DNAH5     NM_001369  exon28   chr5    13864642   13864642   \n",
       "1     9   DNAH5     NM_001369  exon35   chr5    13839514   13839514   \n",
       "2    10   DNAH5     NM_001369  exon71   chr5    13721298   13721298   \n",
       "3    14   DNAH5     NM_001369  exon76   chr5    13708242   13708242   \n",
       "4     2   WSCD2     NM_014653   exon2  chr12   108589770  108589770   \n",
       "5     3   WSCD2     NM_014653   exon3  chr12   108600174  108600174   \n",
       "6    10   WSCD2     NM_014653   exon5  chr12   108618575  108618575   \n",
       "7     8     RB1     NM_000321   exon6  chr13    48923147   48923147   \n",
       "8     1  ANKMY2     NM_020319   exon3   chr7    16666689   16666689   \n",
       "9    10  PCDHA5     NM_018908   exon1   chr5   140203618  140203618   \n",
       "10    7  RAD54B  NM_001205263   exon4   chr8    95416375   95416375   \n",
       "11    3    DDB1     NM_001923   exon6  chr11    61093159   61093159   \n",
       "12    7    TP53     NM_000546   exon5  chr17     7578457    7578457   \n",
       "13    8    TP53     NM_000546   exon6  chr17     7578268    7578268   \n",
       "14    8    TP53     NM_000546   exon7  chr17     7577538    7577538   \n",
       "15    8    TP53     NM_000546   exon7  chr17     7577570    7577570   \n",
       "16    9    TP53     NM_000546   exon5  chr17     7578406    7578406   \n",
       "17   14    TP53     NM_000546   exon6  chr17     7578253    7578253   \n",
       "18   14    TP53     NM_000546   exon5  chr17     7578442    7578442   \n",
       "19   14    TP53     NM_000546   exon8  chr17     8000000    7577121   \n",
       "\n",
       "    Start Forward-Primer  End Forward-Primer     Forward Primer 5'-3'  \\\n",
       "0               13864704            13864723     TGAAGGACTGGCAGGCTTTT   \n",
       "1               13839596            13839615     CCATGGACTTTGAGTGGCTG   \n",
       "2               13721335            13721352       GCCCGCAAGTACATCGTG   \n",
       "3               13708323            13708342     ACCCTCACTGAGCTGAAACT   \n",
       "4              108589697           108589716     GGGAGCCTTGTCTTCCTTCA   \n",
       "5              108600069           108600088     AGTACATCGGCTGCTACCTG   \n",
       "6              108618457           108618476     TCCCAACCAGATCTGACTCC   \n",
       "7               48923119            48923141  GGTGCTAAAAGTTTCTTGGATCA   \n",
       "8               16666758            16666780  TGCAGCATATAAAGGAAAACTCG   \n",
       "9              140203570           140203553       GGCAAGCCCACTCTGTTG   \n",
       "10              95416435            95416455    TGCCACGACCAGATAAGAATC   \n",
       "11              61093221            61093240      CAGTCCTGGTTTGGCAGAC   \n",
       "12               7578530             7578549     CCCTGCCCTCAACAAGATGT   \n",
       "13               7578300             7578319     GGCCTCTGATTCCTCACTGA   \n",
       "14               7577591             7577610     AGGTTGGCTCTGACTGTACC   \n",
       "15               7577591             7577610     AGGTTGGCTCTGACTGTACC   \n",
       "16               7578530             7578549     CCCTGCCCTCAACAAGATGT   \n",
       "17               7578300             7578319     GGCCTCTGATTCCTCACTGA   \n",
       "18               7578530             7578549     CCCTGCCCTCAACAAGATGT   \n",
       "19               7577185             7577205    GGACCTGATTTCCTTACTGCC   \n",
       "\n",
       "    Start Reverse-Primer  End Reverse-Primer    Reverse Primer 5'-3'   \\\n",
       "0               13864569            13864588     ATTCCCCACATCCAGACTGT   \n",
       "1               13839441            13839460      GGTTGGGGAGAAGGGTTCC   \n",
       "2               13721196            13721215     CTTCCCCAAGGCAATGATGG   \n",
       "3               13708154            13708174     TCATGCTTTCCACTTGCCAA   \n",
       "4              108589834           108589853     CTTCACCTGTGTCCCGGAAA   \n",
       "5              108600194           108600210    TGGGAATCTGAGAAGGAAGGG   \n",
       "6              108618605           108618624     AAGTCCACGCATTTGTCCAC   \n",
       "7               48923242            48923259     GCAGAGAATGAGGGAGGAGT   \n",
       "8               16666663            16666682     ACCAGAAAGTGCAGCAAACA   \n",
       "9              140203686           140203712     CGTGGGTTGTCTGTAGAGGT   \n",
       "10              95416312            95416329  TCCTACCTCATTCCCATTACACA   \n",
       "11              61093080            61093099     ACTGCGCTCACCTTGATGAT   \n",
       "12               7578364             7578383     TGCTCACCATCGCTATCTGA   \n",
       "13               7578164             7578183     TTGCAAACCAGACCTCAGGC   \n",
       "14               7577431             7577452       GGGTCAGAGGCAAGCAGA   \n",
       "15               7577431             7577452       GGGTCAGAGGCAAGCAGA   \n",
       "16               7578364             7578383     TGCTCACCATCGCTATCTGA   \n",
       "17               7578164             7578183     TTGCAAACCAGACCTCAGGC   \n",
       "18               7578364             7578383     TGCTCACCATCGCTATCTGA   \n",
       "19               7576997             7577018     GCTTCTTGTCCTGCTTGCTT   \n",
       "\n",
       "    Product size (bp)                                        Comment  \n",
       "0                 155                                            NaN  \n",
       "1                 175                              R-Primer intronic  \n",
       "2                 159                                            NaN  \n",
       "3                 188                              R-Primer intronic  \n",
       "4                 157                                            NaN  \n",
       "5                 169                              R-Primer intronic  \n",
       "6                 168                              F-Primer intronic  \n",
       "7                 158  F-Primer close to Mutation; R-Primer intronic  \n",
       "8                 118                R-Primer very close to mutation  \n",
       "9                 158                                            NaN  \n",
       "10                157                                            NaN  \n",
       "11                169                                            NaN  \n",
       "12                186                                            NaN  \n",
       "13                156                                            NaN  \n",
       "14                178                                            NaN  \n",
       "15                178                                            NaN  \n",
       "16                186                                            NaN  \n",
       "17                156                                            NaN  \n",
       "18                186                                            NaN  \n",
       "19                209                                            NaN  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_file = os.path.join(home, \"Desktop/NEBnextCoop/CoopPrimerList.xlsx\")\n",
    "mut_df = pd.read_excel(sample_file, header=1)\n",
    "mut_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## here the code from primer3 core (in somvarWES)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import primer3\n",
    "import re\n",
    "import pandas as pd\n",
    "from multiprocessing import Pool\n",
    "from functools import partial\n",
    "from script_utils import show_output\n",
    "\n",
    "\n",
    "def file2str(file):\n",
    "    '''\n",
    "    returns a string from a text file\n",
    "    '''\n",
    "\n",
    "    with open(file, 'r') as file:\n",
    "        return file.read().upper().replace('\\n', '')\n",
    "\n",
    "\n",
    "def get_chrom(chrom, chroms_folder='chroms'):\n",
    "    '''\n",
    "    convenience function returning the chromosome sequence without\n",
    "    the header sequences\n",
    "    a folder containing the split chromosomes has to be provided\n",
    "    '''\n",
    "\n",
    "    if 'CHR' in str(chrom).upper():\n",
    "        chrom = chrom.upper().replace('CHR', '')\n",
    "    chrom_file = f\"{chroms_folder}/chr{chrom}.fa\"\n",
    "    chrom_seq = file2str(chrom_file)\n",
    "    chrom_seq = re.sub(r'^>CHR[0-9XY]+', '', chrom_seq)\n",
    "    chrom = {\n",
    "        \"name\": f\"chr{chrom}\",\n",
    "        \"sequence\": chrom_seq,\n",
    "        \"length\": len(chrom_seq)\n",
    "    }\n",
    "    return chrom\n",
    "\n",
    "\n",
    "def mut2insert(mut={}, seq={}, return_none=False):\n",
    "    '''\n",
    "    takes a mutation dictionary of shape:\n",
    "    {\n",
    "        'Chrom': 'chr4,\n",
    "        'Start': 13331,\n",
    "        'End': 13331,\n",
    "        'Ref': '-',\n",
    "        'Alt': 'A'\n",
    "        }\n",
    "    and a seq dictionary of shape:\n",
    "    {\n",
    "        'Chrom': 'chr4,\n",
    "        'Start': 13300,\n",
    "        'End': 123123,\n",
    "        'seq': 'ATTTCTCCCACTCCCACA',\n",
    "        }\n",
    "    and returns the seq with the mutation inserted into the sequence\n",
    "    if mutation location is out of bounds of sequence, only the sequence is returned without editing\n",
    "    '''\n",
    "\n",
    "    if mut['Chr'] != seq['Chr']:\n",
    "        return None if return_none else seq['seq']\n",
    "    if (mut['Start'] < seq['Start']) or (mut['End'] > seq['End']):\n",
    "        return None if return_none else seq['seq']\n",
    "    # case SNP\n",
    "    start = mut['Start'] - seq['Start']\n",
    "    end = mut['End'] - seq['Start']\n",
    "\n",
    "    if mut['Ref'] == \"-\":\n",
    "        upstream = seq['seq'][:start]\n",
    "        downstream = seq['seq'][end:]\n",
    "        mutation = f\"<<+{mut['Alt']}>>\"\n",
    "    else:\n",
    "        upstream = seq['seq'][:start]\n",
    "        downstream = seq['seq'][end + 1:]\n",
    "        if mut['Alt'] == \"-\":\n",
    "            mutation = f\">âˆ†{mut['Ref']}<\"\n",
    "        else:\n",
    "            # SNP\n",
    "            mutation = f\"({mut['Ref']}>{mut['Alt']})\"\n",
    "    return f\"{upstream}{mutation}{downstream}\"\n",
    "\n",
    "\n",
    "def edit_seq(row):\n",
    "    # convert mutation to dict\n",
    "    def mut2dict(row):\n",
    "        # mutation to dict\n",
    "        mut_dict = {\n",
    "            'Chr': row['Chr'],\n",
    "            'Start': row['Start'],\n",
    "            'End': row['End'],\n",
    "            'Ref': row['Ref'],\n",
    "            'Alt': row['Alt']\n",
    "        }\n",
    "        return mut_dict\n",
    "    mut_dict = mut2dict(row)\n",
    "\n",
    "    insert_dict = {\n",
    "        'Chr': row['Chr'],\n",
    "        'Start': row['InsertStart'],\n",
    "        'End': row['InsertEnd'],\n",
    "        'seq': row['InsertSeq']\n",
    "    }\n",
    "\n",
    "    edited_seq = mut2insert(mut=mut_dict, seq=insert_dict, return_none=True)\n",
    "    return edited_seq\n",
    "\n",
    "\n",
    "def compute_primers(row, chrom, config):\n",
    "    '''\n",
    "    row_based primer3 worker\n",
    "    returns the best primer pair for a given position\n",
    "    return value is [fwd_seq, fwd_tmp, rev_seq, rev_tmp, prod_size]\n",
    "    active chromosome sequence is global variable chrom\n",
    "    '''\n",
    "\n",
    "    # load sequence\n",
    "    pos = row['Start']\n",
    "    half_seq = int(config['seq_len'] / 2)\n",
    "    seq_start = pos - half_seq\n",
    "    seq_end = pos + half_seq\n",
    "    seq = chrom['sequence'][seq_start:seq_end]\n",
    "    pad = int(config['mut_pad'] / 2)\n",
    "    half_size = int(config['prod_size_min'] / 2)\n",
    "\n",
    "    # calculate the target_range as offSet from center (half)\n",
    "    offSet = half_size - 20 - pad\n",
    "    target_start = half_seq - offSet\n",
    "    target = [target_start, offSet * 2]\n",
    "    setting = {\n",
    "        'SEQUENCE_ID': 'asdf',\n",
    "        'SEQUENCE_TEMPLATE': seq,\n",
    "        'SEQUENCE_TARGET': target\n",
    "    }\n",
    "    primers = primer3.bindings.designPrimers(setting, config)\n",
    "\n",
    "    # return '--' if nothing was found\n",
    "    if primers['PRIMER_PAIR_NUM_RETURNED'] == 0:\n",
    "        row['fwd_seq'] = row['fwd_tmp'] = row['rev_seq'] = row['rev_tmp'] = row['prod_size'] = '--'\n",
    "        return row\n",
    "\n",
    "    # # get chrom coords\n",
    "    amp_start = seq_start + primers['PRIMER_LEFT_0'][0] + 1\n",
    "    amp_end = seq_start + primers['PRIMER_RIGHT_0'][0] + 1\n",
    "\n",
    "    row['AmpliconRange'] = f\"{chrom['name']}:{amp_start}-{amp_end}\"\n",
    "\n",
    "    insert_start = amp_start + primers['PRIMER_LEFT_0'][1]\n",
    "    insert_end = amp_end - primers['PRIMER_RIGHT_0'][1]\n",
    "    row['InsertRange'] = f\"{chrom['name']}:{insert_start}-{insert_end}\"\n",
    "    row['InsertSize'] = insert_end - insert_start\n",
    "    insert_seq = chrom['sequence'][insert_start - 1: insert_end]\n",
    "    row['InsertSeq'] = insert_seq\n",
    "\n",
    "    mut_dict = {\n",
    "        'Chr': row['Chr'],\n",
    "        'Start': row['Start'],\n",
    "        'End': row['End'],\n",
    "        'Ref': row['Ref'],\n",
    "        'Alt': row['Alt']\n",
    "    }\n",
    "\n",
    "    insert_dict = {\n",
    "        'Chr': row['Chr'],\n",
    "        'Start': insert_start,\n",
    "        'End': insert_end,\n",
    "        'seq': insert_seq\n",
    "    }\n",
    "\n",
    "    row['InsertSeq'] = mut2insert(\n",
    "        mut=mut_dict, seq=insert_dict, return_none=True)\n",
    "    row['offsetL'] = row['Start'] - insert_start\n",
    "    row['offsetR'] = insert_end - row['End']\n",
    "\n",
    "    row['fwdPrimer'] = primers['PRIMER_LEFT_0_SEQUENCE']\n",
    "    row['revPrimer'] = primers['PRIMER_RIGHT_0_SEQUENCE']\n",
    "    row['AmpliconSize'] = primers['PRIMER_RIGHT_0'][0] - \\\n",
    "        primers['PRIMER_LEFT_0'][0] + 1\n",
    "    row['Status'] = 'not established'\n",
    "    row['Note'] = f\"(fwd={int(primers['PRIMER_LEFT_0_TM'] * 10) / 10}|rev={int(primers['PRIMER_RIGHT_0_TM'] * 10) / 10})\"\n",
    "    return row\n",
    "\n",
    "\n",
    "def get_primer_df(mut_df, primer3_config, chroms_folder, chrom):\n",
    "    '''\n",
    "    allocates dfs chrom-wise and controls row-wise computation\n",
    "    '''\n",
    "    show_output(f\"Running primer3 for chrom {chrom}.\", multi=True)\n",
    "    chrom_dict = get_chrom(chrom, chroms_folder)\n",
    "    chr_df = mut_df.query('Chr == @chrom')\n",
    "    primer_df = chr_df.apply(\n",
    "        compute_primers, chrom=chrom_dict, config=primer3_config, axis=1)\n",
    "    show_output(f\"Finished chrom {chrom}.\", multi=True)\n",
    "    return primer_df\n",
    "\n",
    "\n",
    "def run_primer3(mut_df, chroms_folder, pcr_config, primer3_config, threads):\n",
    "\n",
    "    # apply pcr size to primer3_config\n",
    "    primer3_config['PRIMER_PRODUCT_SIZE_RANGE'] = [\n",
    "        pcr_config['prod_size_min'], pcr_config['prod_size_max']]\n",
    "    primer3_config.update(pcr_config)\n",
    "\n",
    "    mut_df.loc[:, 'Chr'] = mut_df['Chr'].astype('str')\n",
    "    mut_df = mut_df.loc[:, ['Chr', 'Start', 'End', 'Ref', 'Alt', 'Gene']]\n",
    "    org_cols = list(mut_df.columns)\n",
    "    df_list = []\n",
    "    # cycle through (formatted) chromosomes\n",
    "    # + load chromosome sequence\n",
    "    # + create primer_df for mutations on that chromosome\n",
    "    # + concat all mutations\n",
    "\n",
    "    # ##### MULTIPROCESSING\n",
    "    chrom_list = mut_df['Chr'].unique()\n",
    "    show_output(f\"Allocating processor pool for {threads} threads.\")\n",
    "    pool = Pool(threads)\n",
    "    df_list = pool.map(partial(get_primer_df, mut_df,\n",
    "                               primer3_config, chroms_folder), chrom_list)\n",
    "\n",
    "    primer_df = pd.concat(df_list, sort=True)\n",
    "    new_cols = [\n",
    "        'fwdPrimer',\n",
    "        'revPrimer',\n",
    "        'Status',\n",
    "        'Note',\n",
    "        'AmpliconRange',\n",
    "        'AmpliconSize',\n",
    "        'InsertRange',\n",
    "        'InsertSize',\n",
    "        'InsertSeq',\n",
    "        'offsetL',\n",
    "        'offsetR'\n",
    "    ]\n",
    "    for col in ['AmpliconSize', 'InsertSize', 'offsetL', 'offsetR']:\n",
    "        primer_df[col] = primer_df[col].fillna(0).astype(int)\n",
    "\n",
    "    primer_df = primer_df[org_cols + new_cols]\n",
    "    return primer_df\n",
    "\n",
    "\n",
    "def primer3_main(i, o, chroms_folder, threads, PCR_config={\n",
    "        'seq_len': 500,\n",
    "        'mut_pad': 5,\n",
    "        'prod_size_min': 120,\n",
    "        'prod_size_max': 220\n",
    "}):\n",
    "    '''\n",
    "    wrapper around run_primer3 that allows for injecting with adjusted PCR and Primer3_configs\n",
    "    and controls input and output\n",
    "    '''\n",
    "\n",
    "    primer3_config = {\n",
    "        'PRIMER_OPT_SIZE': 20,\n",
    "        'PRIMER_PICK_INTERNAL_OLIGO': 0,\n",
    "        'PRIMER_INTERNAL_MAX_SELF_END': 8,\n",
    "        'PRIMER_MIN_SIZE': 18,\n",
    "        'PRIMER_MAX_SIZE': 25,\n",
    "        'PRIMER_OPT_TM': 60.0,\n",
    "        'PRIMER_MIN_TM': 55.0,\n",
    "        'PRIMER_MAX_TM': 65.0,\n",
    "        'PRIMER_MIN_GC': 20.0,\n",
    "        'PRIMER_MAX_GC': 80.0,\n",
    "        'PRIMER_MAX_POLY_X': 100,\n",
    "        'PRIMER_INTERNAL_MAX_POLY_X': 100,\n",
    "        'PRIMER_SALT_MONOVALENT': 50.0,\n",
    "        'PRIMER_DNA_CONC': 50.0,\n",
    "        'PRIMER_MAX_NS_ACCEPTED': 0,\n",
    "        'PRIMER_MAX_SELF_ANY': 12,\n",
    "        'PRIMER_MAX_SELF_END': 8,\n",
    "        'PRIMER_PAIR_MAX_COMPL_ANY': 12,\n",
    "        'PRIMER_PAIR_MAX_COMPL_END': 8,\n",
    "    }\n",
    "    # #### LOAD file ###################\n",
    "    show_output(f\"Loading {i} for primer3 computation. \", end='')\n",
    "    filter1_df = pd.read_csv(i, sep='\\t')\n",
    "    show_output(f\"{len(filter1_df.index)} mutations found.\", time=False)\n",
    "\n",
    "    # #### run primer3 #################\n",
    "    primer_df = run_primer3(\n",
    "        filter1_df,\n",
    "        chroms_folder,\n",
    "        pcr_config=PCR_config,\n",
    "        primer3_config=primer3_config,\n",
    "        threads=threads\n",
    "    )\n",
    "\n",
    "    # ###### write to file #############\n",
    "    primer_df.to_csv(o, sep='\\t', index=False)\n",
    "    show_output(f\"Writing primer list to {o}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-18T08:46:40.175092Z",
     "start_time": "2019-09-18T08:46:40.141444Z"
    }
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'PCR_config' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-12-179e0013d16c>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m    109\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    110\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 111\u001b[0;31m \u001b[0;32mdef\u001b[0m \u001b[0mrun_primer3\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmut_df\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgenome_split_folder\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m''\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpcr_config\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mPCR_config\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprimer3_config\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mPrimer3_config\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    112\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    113\u001b[0m     \u001b[0;31m# apply pcr size to primer3_config\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'PCR_config' is not defined"
     ]
    }
   ],
   "source": [
    "PCR_config = {\n",
    "    'seq_len': 500,\n",
    "    'mut_pad': 5,\n",
    "    'prod_size_min': 120,\n",
    "    'prod_size_max': 220\n",
    "}\n",
    "\n",
    "Primer3_config = {\n",
    "        'PRIMER_OPT_SIZE': 20,\n",
    "        'PRIMER_PICK_INTERNAL_OLIGO': 0,\n",
    "        'PRIMER_INTERNAL_MAX_SELF_END': 8,\n",
    "        'PRIMER_MIN_SIZE': 18,\n",
    "        'PRIMER_MAX_SIZE': 25,\n",
    "        'PRIMER_OPT_TM': 60.0,\n",
    "        'PRIMER_MIN_TM': 57.0,\n",
    "        'PRIMER_MAX_TM': 63.0,\n",
    "        'PRIMER_MIN_GC': 20.0,\n",
    "        'PRIMER_MAX_GC': 80.0,\n",
    "        'PRIMER_MAX_POLY_X': 100,\n",
    "        'PRIMER_INTERNAL_MAX_POLY_X': 100,\n",
    "        'PRIMER_SALT_MONOVALENT': 50.0,\n",
    "        'PRIMER_DNA_CONC': 50.0,\n",
    "        'PRIMER_MAX_NS_ACCEPTED': 0,\n",
    "        'PRIMER_MAX_SELF_ANY': 12,\n",
    "        'PRIMER_MAX_SELF_END': 8,\n",
    "        'PRIMER_PAIR_MAX_COMPL_ANY': 12,\n",
    "        'PRIMER_PAIR_MAX_COMPL_END': 8,\n",
    "    }\n",
    "\n",
    "\n",
    "def file2str(file):\n",
    "    '''\n",
    "    returns a string from a text file\n",
    "    '''\n",
    "    \n",
    "    with open(file, 'r') as file:\n",
    "        return file.read().upper().replace('\\n', '')\n",
    "    \n",
    "def get_primer_df(chrom_seq, primer_config, row):\n",
    "    '''\n",
    "    returns the best primer pair for a given position\n",
    "    return value is [fwd_seq, fwd_tmp, rev_seq, rev_tmp, prod_size]\n",
    "    active chromosome sequence is global variable chrom\n",
    "    '''\n",
    "    \n",
    "    # load sequence\n",
    "    pos = row['Pos']\n",
    "    half_seq = int(config['seq_len']/ 2)\n",
    "    seq_start = pos - half_seq + 6  # offSet 6 because of \n",
    "    seq_end = pos + half_seq + 6\n",
    "    seq = chrom_seq[seq_start:seq_end]\n",
    "    pad = int(config['mut_pad'] / 2)\n",
    "    half_size = int(config['prod_size_min'] / 2)\n",
    "    \n",
    "    # calculate the target_range as offSet from center (half)\n",
    "    offSet = half_size - 20 - pad\n",
    "    target_start = half_seq - offSet\n",
    "    target = [target_start, offSet * 2]\n",
    "    # print(target)\n",
    "    setting = {\n",
    "        'SEQUENCE_ID': 'asdf',\n",
    "        'SEQUENCE_TEMPLATE': seq,\n",
    "        'SEQUENCE_TARGET': target\n",
    "    }\n",
    "    primers = primer3.bindings.designPrimers(setting, primer_config)\n",
    "    # print(primers)\n",
    "    \n",
    "    # return '--' if nothing was found\n",
    "    if primers['PRIMER_PAIR_NUM_RETURNED'] == 0:\n",
    "        row['fwd_seq'] = row['fwd_tmp'] = row['rev_seq'] = row['rev_tmp'] = row['prod_size'] = '--'\n",
    "        return row\n",
    "    \n",
    "    \n",
    "    prod_len = primers['PRIMER_RIGHT_0'][0] - primers['PRIMER_LEFT_0'][0]\n",
    "    row['fwd_seq'] = primers['PRIMER_LEFT_0_SEQUENCE']\n",
    "    row['fwd_tmp'] = primers['PRIMER_LEFT_0_TM']\n",
    "    row['rev_seq'] = primers['PRIMER_RIGHT_0_SEQUENCE']\n",
    "    row['rev_tmp'] = primers['PRIMER_RIGHT_0_TM']\n",
    "    row['prod_size'] = prod_len\n",
    "    return row\n",
    "\n",
    "\n",
    "def make_primer_list(mut_file, PCR_config, P3_config, chrom_column='Chr', pos_column='Pos', keep_cols='all'):\n",
    "    \n",
    "    \n",
    "    # apply pcr size to config\n",
    "    P3_config['PRIMER_PRODUCT_SIZE_RANGE'] = [PCR_config['prod_size_min'],PCR_config['prod_size_max']]\n",
    "    # get the mutation list\n",
    "    if os.path.splitext(mut_file)[1] in ['.xlsx', '.xls']:\n",
    "        mut_list = pd.read_excel(mut_file, header=0)\n",
    "    else:\n",
    "        mut_list = pd.read_csv(mut_file, header=0)\n",
    "        \n",
    "    # replace column names for chrom and pos\n",
    "    mut_list = mut_list.rename(columns={chrom_column: 'Chr', pos_column: 'Pos'})\n",
    "    mut_list['Chr'] = mut_list['Chr'].astype('str')\n",
    "    org_cols = list(mut_list.columns)\n",
    "    df_list = []\n",
    "    # cycle through (formatted) chromosomes\n",
    "    # + load chromosome sequence\n",
    "    # + create primer_df for mutations on that chromosome\n",
    "    # + concat all mutations\n",
    "    for chrom in [str(chrom).lower().replace('chr','').upper() for chrom in mut_list['Chr'].unique()]:\n",
    "        chrom_seq = file2str(f'hg38/chr{chrom}.fa')\n",
    "        print(chrom)\n",
    "        df_list.append(mut_list.query('Chr == @chrom').apply(partial(get_primer_df, chrom_seq, P3_config), axis=1))\n",
    "    primer_df = pd.concat(df_list, sort=True)\n",
    "    if keep_cols == 'all':\n",
    "        primer_df = primer_df[org_cols[:5] + ['fwd_seq', 'fwd_tmp', 'rev_seq', 'rev_tmp', 'prod_size'] + org_cols[5:]]\n",
    "    else:\n",
    "        primer_df = primer_df[org_cols[:keep_cols] + ['fwd_seq', 'fwd_tmp', 'rev_seq', 'rev_tmp', 'prod_size']]\n",
    "        \n",
    "    primer_df.to_csv(f\"{os.path.splitext(mut_file)[0]}-primers.csv\")\n",
    "    return primer_df\n",
    "\n",
    "\n",
    "\n",
    "def run_primer3(mut_df, genome_split_folder='', pcr_config=PCR_config, primer3_config=Primer3_config):\n",
    "\n",
    "    # apply pcr size to primer3_config\n",
    "    primer3_config['PRIMER_PRODUCT_SIZE_RANGE'] = [pcr_config['prod_size_min'], pcr_config['prod_size_max']]\n",
    "    primer3_config.update(pcr_config)\n",
    "\n",
    "    mut_df.loc[:, 'Chr'] = mut_df['Chr'].astype('str')\n",
    "    org_cols = list(mut_df.columns)\n",
    "    df_list = []\n",
    "    # cycle through (formatted) chromosomes\n",
    "    # + load chromosome sequence\n",
    "    # + create primer_df for mutations on that chromosome\n",
    "    # + concat all mutations\n",
    "    for chrom in mut_df['Chr'].unique():\n",
    "        chrom_seq = file2str(f'{genome_split_folder}/{chrom}.fa')\n",
    "        chr_df = mut_df.query('Chr == @chrom')\n",
    "        primer_df = chr_df.apply(partial(get_primer_df, chrom_seq, primer3_config), axis=1)\n",
    "        df_list.append(primer_df)\n",
    "    primer_df = pd.concat(df_list, sort=True)\n",
    "    clinscore_index = list(mut_df.columns).index('Clin_score') + 1\n",
    "    primer_df = primer_df[org_cols[:clinscore_index] + ['PrimerFwd', 'PrimerFwd_Tmp', 'PrimerRev', 'PrimerRev_Tmp', 'ProductSize', 'MutLocation'] + org_cols[clinscore_index:]]\n",
    "    return primer_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-18T08:46:52.330256Z",
     "start_time": "2019-09-18T08:46:41.468838Z"
    }
   },
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'hg38/chr9.fa'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-10-a7e6da6d9e8a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mprimer_list\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmake_primer_list\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'selected_mutations.xlsx'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconfig\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprimer_config\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mchrom_column\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'CHROM'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpos_column\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'hg38'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkeep_cols\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m5\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mprimer_list\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-9-2ca9111542e5>\u001b[0m in \u001b[0;36mmake_primer_list\u001b[0;34m(mut_file, PCR_config, P3_config, chrom_column, pos_column, keep_cols)\u001b[0m\n\u001b[1;32m    103\u001b[0m     \u001b[0;31m# + concat all mutations\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    104\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mchrom\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchrom\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlower\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'chr'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m''\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mchrom\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mmut_list\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Chr'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 105\u001b[0;31m         \u001b[0mchrom_seq\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfile2str\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'hg38/chr{chrom}.fa'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    106\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchrom\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    107\u001b[0m         \u001b[0mdf_list\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmut_list\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mquery\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Chr == @chrom'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpartial\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mget_primer_df\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mchrom_seq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mP3_config\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-9-2ca9111542e5>\u001b[0m in \u001b[0;36mfile2str\u001b[0;34m(file)\u001b[0m\n\u001b[1;32m     35\u001b[0m     '''\n\u001b[1;32m     36\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 37\u001b[0;31m     \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'r'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mfile\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     38\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mfile\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'\\n'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     39\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'hg38/chr9.fa'"
     ]
    }
   ],
   "source": [
    "primer_list = make_primer_list('selected_mutations.xlsx', config, primer_config, chrom_column='CHROM', pos_column='hg38', keep_cols=5)\n",
    "primer_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-11T07:19:48.251789Z",
     "start_time": "2019-09-11T07:19:48.147657Z"
    }
   },
   "outputs": [],
   "source": [
    "mut_list = pd.read_excel('selected_mutations.xlsx', header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-11T07:20:20.988392Z",
     "start_time": "2019-09-11T07:20:20.965174Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UPN</th>\n",
       "      <th>Gene_Name</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>hg19</th>\n",
       "      <th>hg38</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>HGVS.c</th>\n",
       "      <th>HGVS.p</th>\n",
       "      <th>Feature_ID</th>\n",
       "      <th>SDP</th>\n",
       "      <th>RD</th>\n",
       "      <th>AD</th>\n",
       "      <th>dbSNP</th>\n",
       "      <th>Validated</th>\n",
       "      <th>Method</th>\n",
       "      <th>Project</th>\n",
       "      <th>Screening VAF</th>\n",
       "      <th>Validation VAF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      <td>UPN355</td>\n",
       "      <td>ATRX</td>\n",
       "      <td>X</td>\n",
       "      <td>76778798</td>\n",
       "      <td>77523320</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>c.6781A&gt;T</td>\n",
       "      <td>p.(Lys2261*)</td>\n",
       "      <td>ENST00000373344</td>\n",
       "      <td>90</td>\n",
       "      <td>79</td>\n",
       "      <td>10</td>\n",
       "      <td>.</td>\n",
       "      <td>falsified</td>\n",
       "      <td>amplicon based</td>\n",
       "      <td>Non-Cancer</td>\n",
       "      <td>0.1111</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      <td>UPN312</td>\n",
       "      <td>BCOR</td>\n",
       "      <td>X</td>\n",
       "      <td>39933036</td>\n",
       "      <td>40073783</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>c.1563T&gt;A</td>\n",
       "      <td>p.(Asn521Lys)</td>\n",
       "      <td>ENST00000342274</td>\n",
       "      <td>3924</td>\n",
       "      <td>1937</td>\n",
       "      <td>1748</td>\n",
       "      <td>COSM5367252;COSM5367253</td>\n",
       "      <td>validated</td>\n",
       "      <td>amplicon based</td>\n",
       "      <td>Non-Cancer</td>\n",
       "      <td>0.4455</td>\n",
       "      <td>0.4706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      <td>UPN192</td>\n",
       "      <td>BCORL1</td>\n",
       "      <td>X</td>\n",
       "      <td>129147830</td>\n",
       "      <td>130013854</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>c.1082C&gt;T</td>\n",
       "      <td>p.(Thr361Ile)</td>\n",
       "      <td>ENST00000218147</td>\n",
       "      <td>433</td>\n",
       "      <td>353</td>\n",
       "      <td>15</td>\n",
       "      <td>.</td>\n",
       "      <td>falsified</td>\n",
       "      <td>amplicon based</td>\n",
       "      <td>Non-Cancer</td>\n",
       "      <td>0.0346</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>109</td>\n",
       "      <td>UPN287</td>\n",
       "      <td>KDM6A</td>\n",
       "      <td>X</td>\n",
       "      <td>44970628 C</td>\n",
       "      <td>45111383</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>missense_variant, splice_region_variant</td>\n",
       "      <td>c.4178C&gt;T</td>\n",
       "      <td>p.(Ala1393Val)</td>\n",
       "      <td>ENST00000377967</td>\n",
       "      <td>319</td>\n",
       "      <td>281</td>\n",
       "      <td>11</td>\n",
       "      <td>.</td>\n",
       "      <td>falsified</td>\n",
       "      <td>amplicon based</td>\n",
       "      <td>Non-Cancer</td>\n",
       "      <td>0.0345</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>115</td>\n",
       "      <td>UPN332</td>\n",
       "      <td>PHF6</td>\n",
       "      <td>X</td>\n",
       "      <td>133527533 G</td>\n",
       "      <td>134393503</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>missense_variant, splice_region_variant</td>\n",
       "      <td>c.243G&gt;A</td>\n",
       "      <td>p.(Met81Ile)</td>\n",
       "      <td>ENST00000332070</td>\n",
       "      <td>2020</td>\n",
       "      <td>1820</td>\n",
       "      <td>109</td>\n",
       "      <td>.</td>\n",
       "      <td>validated</td>\n",
       "      <td>amplicon based</td>\n",
       "      <td>Non-Cancer</td>\n",
       "      <td>0.0540</td>\n",
       "      <td>0.0425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>137</td>\n",
       "      <td>UPN194</td>\n",
       "      <td>STAG2</td>\n",
       "      <td>X</td>\n",
       "      <td>123156506 A</td>\n",
       "      <td>124022656</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AT</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>c.33dup</td>\n",
       "      <td>p.(Asn12*)</td>\n",
       "      <td>ENST00000218089</td>\n",
       "      <td>2510</td>\n",
       "      <td>2118</td>\n",
       "      <td>131</td>\n",
       "      <td>.</td>\n",
       "      <td>validated</td>\n",
       "      <td>amplicon based</td>\n",
       "      <td>Non-Cancer</td>\n",
       "      <td>0.0522</td>\n",
       "      <td>0.0323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>197</td>\n",
       "      <td>UPN095</td>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>X</td>\n",
       "      <td>15833949</td>\n",
       "      <td>15815826</td>\n",
       "      <td>TC</td>\n",
       "      <td>T</td>\n",
       "      <td>frameshift_truncation</td>\n",
       "      <td>c.709del</td>\n",
       "      <td>p.(Leu237fs)</td>\n",
       "      <td>ENST00000307771 915</td>\n",
       "      <td>819 40</td>\n",
       "      <td>.</td>\n",
       "      <td>validated</td>\n",
       "      <td>amplicon based Non-Cancer 0,0437</td>\n",
       "      <td>0,0451</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>198</td>\n",
       "      <td>UPN260</td>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>X</td>\n",
       "      <td>15841025</td>\n",
       "      <td>15822902</td>\n",
       "      <td>AC</td>\n",
       "      <td>A</td>\n",
       "      <td>frameshift_truncation</td>\n",
       "      <td>c.1110del</td>\n",
       "      <td>p.(His370Glnfs*?)</td>\n",
       "      <td>ENST00000307771 2931</td>\n",
       "      <td>2760 109</td>\n",
       "      <td>.</td>\n",
       "      <td>validated</td>\n",
       "      <td>amplicon based Non-Cancer 0,0372</td>\n",
       "      <td>0,0119</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>199</td>\n",
       "      <td>UPN355</td>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>X</td>\n",
       "      <td>15821864</td>\n",
       "      <td>15803741</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>c.257T&gt;C</td>\n",
       "      <td>p.(Phe86Ser)</td>\n",
       "      <td>ENST00000307771 303</td>\n",
       "      <td>281 11</td>\n",
       "      <td>.</td>\n",
       "      <td>falsified</td>\n",
       "      <td>amplicon based Non-Cancer 0,0363</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>200</td>\n",
       "      <td>UPN415</td>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>X</td>\n",
       "      <td>15821911</td>\n",
       "      <td>15803788</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>c.304G&gt;A</td>\n",
       "      <td>p.(Glu102Lys)</td>\n",
       "      <td>ENST00000307771 670</td>\n",
       "      <td>611 20</td>\n",
       "      <td>.</td>\n",
       "      <td>falsified</td>\n",
       "      <td>amplicon based     Cancer 0,0299</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        UPN Gene_Name CHROM         hg19       hg38  REF ALT  \\\n",
       "16   UPN355      ATRX     X     76778798   77523320    T   A   \n",
       "17   UPN312      BCOR     X     39933036   40073783    A   T   \n",
       "18   UPN192    BCORL1     X    129147830  130013854    C   T   \n",
       "109  UPN287     KDM6A     X   44970628 C   45111383  NaN   T   \n",
       "115  UPN332      PHF6     X  133527533 G  134393503  NaN   A   \n",
       "137  UPN194     STAG2     X  123156506 A  124022656  NaN  AT   \n",
       "197  UPN095     ZRSR2     X     15833949   15815826   TC   T   \n",
       "198  UPN260     ZRSR2     X     15841025   15822902   AC   A   \n",
       "199  UPN355     ZRSR2     X     15821864   15803741    T   C   \n",
       "200  UPN415     ZRSR2     X     15821911   15803788    G   A   \n",
       "\n",
       "                                  Annotation     HGVS.c             HGVS.p  \\\n",
       "16                               stop_gained  c.6781A>T       p.(Lys2261*)   \n",
       "17                          missense_variant  c.1563T>A      p.(Asn521Lys)   \n",
       "18                          missense_variant  c.1082C>T      p.(Thr361Ile)   \n",
       "109  missense_variant, splice_region_variant  c.4178C>T     p.(Ala1393Val)   \n",
       "115  missense_variant, splice_region_variant   c.243G>A       p.(Met81Ile)   \n",
       "137                              stop_gained    c.33dup         p.(Asn12*)   \n",
       "197                    frameshift_truncation   c.709del       p.(Leu237fs)   \n",
       "198                    frameshift_truncation  c.1110del  p.(His370Glnfs*?)   \n",
       "199                         missense_variant   c.257T>C       p.(Phe86Ser)   \n",
       "200                         missense_variant   c.304G>A      p.(Glu102Lys)   \n",
       "\n",
       "               Feature_ID       SDP    RD         AD  \\\n",
       "16        ENST00000373344        90    79         10   \n",
       "17        ENST00000342274      3924  1937       1748   \n",
       "18        ENST00000218147       433   353         15   \n",
       "109       ENST00000377967       319   281         11   \n",
       "115       ENST00000332070      2020  1820        109   \n",
       "137       ENST00000218089      2510  2118        131   \n",
       "197   ENST00000307771 915    819 40     .  validated   \n",
       "198  ENST00000307771 2931  2760 109     .  validated   \n",
       "199   ENST00000307771 303    281 11     .  falsified   \n",
       "200   ENST00000307771 670    611 20     .  falsified   \n",
       "\n",
       "                                dbSNP  Validated          Method     Project  \\\n",
       "16                                  .  falsified  amplicon based  Non-Cancer   \n",
       "17            COSM5367252;COSM5367253  validated  amplicon based  Non-Cancer   \n",
       "18                                  .  falsified  amplicon based  Non-Cancer   \n",
       "109                                 .  falsified  amplicon based  Non-Cancer   \n",
       "115                                 .  validated  amplicon based  Non-Cancer   \n",
       "137                                 .  validated  amplicon based  Non-Cancer   \n",
       "197  amplicon based Non-Cancer 0,0437     0,0451             NaN         NaN   \n",
       "198  amplicon based Non-Cancer 0,0372     0,0119             NaN         NaN   \n",
       "199  amplicon based Non-Cancer 0,0363          .             NaN         NaN   \n",
       "200  amplicon based     Cancer 0,0299          .             NaN         NaN   \n",
       "\n",
       "     Screening VAF Validation VAF  \n",
       "16          0.1111              .  \n",
       "17          0.4455         0.4706  \n",
       "18          0.0346              .  \n",
       "109         0.0345              .  \n",
       "115         0.0540         0.0425  \n",
       "137         0.0522         0.0323  \n",
       "197            NaN            NaN  \n",
       "198            NaN            NaN  \n",
       "199            NaN            NaN  \n",
       "200            NaN            NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut_list.query('CHROM == \"X\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import math\n",
    "test = 1.4\n",
    "math.floor(test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
