{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-18T08:46:39.109669Z",
     "start_time": "2019-09-18T08:46:39.075118Z"
    }
   },
   "outputs": [],
   "source": [
    "# some sensible settings for better output\n",
    "from IPython.display import display\n",
    "pd.set_option('display.max_columns', None)\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "InteractiveShell.ast_node_interactivity = \"all\"\n",
    "pd.set_option('max_colwidth', 200)\n",
    "\n",
    "\n",
    "# get the code\n",
    "import sys\n",
    "sys.path.append('../code')\n",
    "from script_utils import show_output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load mutation list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_df = pd.read_excel('../testdata/mutations.xlsx', engine=\"openpyxl\", sheet_name=\"mutations\")\n",
    "# test_df = pd.read_csv('../testdata/mutations.csv', sep='\\t')\n",
    "test_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## here the code from primer3 core"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import primer3\n",
    "import re\n",
    "import pandas as pd\n",
    "from multiprocessing import Pool\n",
    "from functools import partial\n",
    "\n",
    "\n",
    "def mut2insert(mut={}, seq={}, return_none=False):\n",
    "    \"\"\"\n",
    "    takes a mutation dictionary of shape:\n",
    "    {\n",
    "        'Chrom': 'chr4,\n",
    "        'Start': 13331,\n",
    "        'End': 13331,\n",
    "        'Ref': '-',\n",
    "        'Alt': 'A'\n",
    "        }\n",
    "    and a seq dictionary of shape:\n",
    "    {\n",
    "        'Chrom': 'chr4,\n",
    "        'Start': 13300,\n",
    "        'End': 123123,\n",
    "        'seq': 'ATTTCTCCCACTCCCACA',\n",
    "        }\n",
    "    and returns the seq with the mutation inserted into the sequence\n",
    "    if mutation location is out of bounds of sequence, only the sequence is returned without editing\n",
    "    \"\"\"\n",
    "\n",
    "    if mut[\"Chr\"] != seq[\"Chr\"]:\n",
    "        return None if return_none else seq[\"seq\"]\n",
    "    if (mut[\"Start\"] < seq[\"Start\"]) or (mut[\"End\"] > seq[\"End\"]):\n",
    "        return None if return_none else seq[\"seq\"]\n",
    "    bases = [\"A\", \"C\", \"G\", \"T\"]\n",
    "    # case SNP\n",
    "    start = mut[\"Start\"] - seq[\"Start\"]\n",
    "    end = mut[\"End\"] - seq[\"Start\"]\n",
    "\n",
    "    if mut[\"Ref\"] == \"-\":\n",
    "        upstream = seq[\"seq\"][:start]\n",
    "        downstream = seq[\"seq\"][end:]\n",
    "        mutation = f\"<<+{mut['Alt']}>>\"\n",
    "    else:\n",
    "        upstream = seq[\"seq\"][:start]\n",
    "        downstream = seq[\"seq\"][end + 1 :]\n",
    "        if mut[\"Alt\"] == \"-\":\n",
    "            mutation = f\">âˆ†{mut['Ref']}<\"\n",
    "        else:\n",
    "            # SNP\n",
    "            mutation = f\"({mut['Ref']}>{mut['Alt']})\"\n",
    "    return f\"{upstream}{mutation}{downstream}\"\n",
    "\n",
    "\n",
    "def edit_seq(row):\n",
    "    # convert mutation to dict\n",
    "    def mut2dict(row):\n",
    "        # mutation to dict\n",
    "        mut_dict = {\n",
    "            \"Chr\": row[\"Chr\"],\n",
    "            \"Start\": row[\"Start\"],\n",
    "            \"End\": row[\"End\"],\n",
    "            \"Ref\": row[\"Ref\"],\n",
    "            \"Alt\": row[\"Alt\"],\n",
    "        }\n",
    "        return mut_dict\n",
    "\n",
    "    mut_dict = mut2dict(row)\n",
    "\n",
    "    insert_dict = {\n",
    "        \"Chr\": row[\"Chr\"],\n",
    "        \"Start\": row[\"InsertStart\"],\n",
    "        \"End\": row[\"InsertEnd\"],\n",
    "        \"seq\": row[\"InsertSeq\"],\n",
    "    }\n",
    "\n",
    "    edited_seq = mut2insert(mut=mut_dict, seq=insert_dict, return_none=True)\n",
    "    return edited_seq\n",
    "\n",
    "\n",
    "def compute_primers(row, chrom, config):\n",
    "    \"\"\"\n",
    "    row_based primer3 worker\n",
    "    returns the best primer pair for a given position\n",
    "    return value is [fwd_seq, fwd_tmp, rev_seq, rev_tmp, prod_size]\n",
    "    active chromosome sequence is global variable chrom\n",
    "    \"\"\"\n",
    "\n",
    "    # load sequence\n",
    "    pos = row[\"Start\"]\n",
    "    half_seq = int(config[\"seq_len\"] / 2)\n",
    "    seq_start = pos - half_seq\n",
    "    seq_end = pos + half_seq\n",
    "    seq = chrom[\"sequence\"][seq_start:seq_end]\n",
    "    pad = int(config[\"center_offSet\"] / 2)\n",
    "    half_size = int(config[\"prod_size_min\"] / 2)\n",
    "\n",
    "    # calculate the target_range as offSet from center (half)\n",
    "    offSet = half_size - 20 - pad\n",
    "    target_start = half_seq - offSet\n",
    "    target = [target_start, offSet * 2]\n",
    "    setting = {\n",
    "        \"SEQUENCE_ID\": \"asdf\",\n",
    "        \"SEQUENCE_TEMPLATE\": seq,\n",
    "        \"SEQUENCE_TARGET\": target,\n",
    "    }\n",
    "    primers = primer3.bindings.designPrimers(setting, config)\n",
    "\n",
    "    # return '--' if nothing was found\n",
    "    if primers[\"PRIMER_PAIR_NUM_RETURNED\"] == 0:\n",
    "        row[\"fwd_seq\"] = row[\"fwd_tmp\"] = row[\"rev_seq\"] = row[\"rev_tmp\"] = row[\n",
    "            \"prod_size\"\n",
    "        ] = \"--\"\n",
    "        return row\n",
    "\n",
    "    # # get chrom coords\n",
    "    amp_start = seq_start + primers[\"PRIMER_LEFT_0\"][0] + 1\n",
    "    amp_end = seq_start + primers[\"PRIMER_RIGHT_0\"][0] + 1\n",
    "\n",
    "    row[\"AmpliconRange\"] = f\"{chrom['name']}:{amp_start}-{amp_end}\"\n",
    "\n",
    "    insert_start = amp_start + primers[\"PRIMER_LEFT_0\"][1]\n",
    "    insert_end = amp_end - primers[\"PRIMER_RIGHT_0\"][1]\n",
    "    row[\"InsertRange\"] = f\"{chrom['name']}:{insert_start}-{insert_end}\"\n",
    "    row[\"InsertSize\"] = insert_end - insert_start\n",
    "    insert_seq = chrom[\"sequence\"][insert_start - 1 : insert_end]\n",
    "    row[\"InsertSeq\"] = insert_seq\n",
    "\n",
    "    mut_dict = {\n",
    "        \"Chr\": row[\"Chr\"],\n",
    "        \"Start\": row[\"Start\"],\n",
    "        \"End\": row[\"End\"],\n",
    "        \"Ref\": row[\"Ref\"],\n",
    "        \"Alt\": row[\"Alt\"],\n",
    "    }\n",
    "\n",
    "    insert_dict = {\n",
    "        \"Chr\": row[\"Chr\"],\n",
    "        \"Start\": insert_start,\n",
    "        \"End\": insert_end,\n",
    "        \"seq\": insert_seq,\n",
    "    }\n",
    "\n",
    "    row[\"InsertSeq\"] = mut2insert(mut=mut_dict, seq=insert_dict, return_none=True)\n",
    "    row[\"offsetL\"] = row[\"Start\"] - insert_start\n",
    "    row[\"offsetR\"] = insert_end - row[\"End\"]\n",
    "\n",
    "    row[\"fwdPrimer\"] = primers[\"PRIMER_LEFT_0_SEQUENCE\"]\n",
    "    row[\"revPrimer\"] = primers[\"PRIMER_RIGHT_0_SEQUENCE\"]\n",
    "    row[\"AmpliconSize\"] = primers[\"PRIMER_RIGHT_0\"][0] - primers[\"PRIMER_LEFT_0\"][0] + 1\n",
    "    row[\"Status\"] = \"not established\"\n",
    "    row[\n",
    "        \"Temp\"\n",
    "    ] = f\"(fwd={int(primers['PRIMER_LEFT_0_TM'] * 10) / 10}|rev={int(primers['PRIMER_RIGHT_0_TM'] * 10) / 10})\"\n",
    "    return row\n",
    "\n",
    "\n",
    "def get_primer_df(mut_df, primer3_config, chroms_folder, chrom):\n",
    "    \"\"\"\n",
    "    allocates dfs chrom-wise and controls row-wise computation\n",
    "    \"\"\"\n",
    "    show_output(f\"Running primer3 for chromosome {chrom}.\", multi=True)\n",
    "    chrom_dict = get_chrom(chrom, chroms_folder)\n",
    "    chr_df = mut_df.query(\"Chr == @chrom\")\n",
    "    primer_df = chr_df.apply(\n",
    "        compute_primers, chrom=chrom_dict, config=primer3_config, axis=1\n",
    "    )\n",
    "    show_output(f\"Finished chrom {chrom}.\", multi=True)\n",
    "    return primer_df\n",
    "\n",
    "\n",
    "def run_primer3(\n",
    "    mut_df,\n",
    "    chroms_folder=\".\",\n",
    "    pcr_config={},  # use defaults defined at top\n",
    "    primer3_config={},  # use defaults defined at top\n",
    "    threads=1,\n",
    "):\n",
    "    \"\"\"\n",
    "    input is df with columns 'Chr', 'Start', 'End', 'Ref', 'Alt' with optional id columns (everything left of Chr)\n",
    "    output is id columns + 'Chr', 'Start', 'End', 'Ref', 'Alt' + primer cols\n",
    "    primer cols:\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    # apply pcr size to primer3_config\n",
    "    primer3_config[\"PRIMER_PRODUCT_SIZE_RANGE\"] = [\n",
    "        pcr_config[\"prod_size_min\"],\n",
    "        pcr_config[\"prod_size_max\"],\n",
    "    ]\n",
    "    primer3_config.update(pcr_config)\n",
    "\n",
    "    mut_df.loc[:, \"Chr\"] = mut_df[\"Chr\"].astype(\"str\")\n",
    "\n",
    "    # COLS\n",
    "    base_cols = [\"Chr\", \"Start\", \"End\", \"Ref\", \"Alt\"]\n",
    "    # keep possible columns left of Chr\n",
    "    # save the id columns into org_df for later merge\n",
    "    keep_cols = list(mut_df.columns[: list(mut_df.columns).index(\"Chr\")])\n",
    "\n",
    "    keep_df = mut_df.loc[:, keep_cols + base_cols]\n",
    "\n",
    "    mut_df = mut_df.loc[:, base_cols]\n",
    "\n",
    "    df_list = []\n",
    "    # cycle through (formatted) chromosomes\n",
    "    # + load chromosome sequence\n",
    "    # + create primer_df for mutations on that chromosome\n",
    "    # + concat all mutations\n",
    "\n",
    "    # ##### MULTIPROCESSING\n",
    "    chrom_list = mut_df[\"Chr\"].unique()\n",
    "    show_output(f\"Allocating processor pool for {threads} threads.\")\n",
    "    pool = Pool(threads)\n",
    "    df_list = pool.map(\n",
    "        partial(get_primer_df, mut_df, primer3_config, chroms_folder), chrom_list\n",
    "    )\n",
    "\n",
    "    primer_df = pd.concat(df_list, sort=True)\n",
    "    new_cols = [\n",
    "        \"fwdPrimer\",\n",
    "        \"revPrimer\",\n",
    "        \"Status\",\n",
    "        \"Temp\",\n",
    "        \"AmpliconRange\",\n",
    "        \"AmpliconSize\",\n",
    "        \"InsertRange\",\n",
    "        \"InsertSize\",\n",
    "        \"InsertSeq\",\n",
    "        \"offsetL\",\n",
    "        \"offsetR\",\n",
    "    ]\n",
    "    for col in [\"AmpliconSize\", \"InsertSize\", \"offsetL\", \"offsetR\"]:\n",
    "        primer_df[col] = primer_df[col].fillna(0).astype(int)\n",
    "\n",
    "    primer_df = keep_df.merge(primer_df[base_cols + new_cols])\n",
    "\n",
    "    return primer_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set the targeted product size and play around with the configs\n",
    "\n",
    "PCR_config = {\n",
    "    'seq_len': 500,\n",
    "    'center_offSet': 5,\n",
    "    'prod_size_min': 120,\n",
    "    'prod_size_max': 220\n",
    "}\n",
    "\n",
    "primer3_config = {\n",
    "        'PRIMER_OPT_SIZE': 20,\n",
    "        'PRIMER_MIN_SIZE': 18,\n",
    "        'PRIMER_MAX_SIZE': 25,\n",
    "        'PRIMER_OPT_TM': 60.0,\n",
    "        'PRIMER_MIN_TM': 55.0,\n",
    "        'PRIMER_MAX_TM': 65.0,\n",
    "        'PRIMER_MIN_GC': 20.0,\n",
    "        'PRIMER_MAX_GC': 80.0,\n",
    "        'PRIMER_PICK_INTERNAL_OLIGO': 0,\n",
    "        'PRIMER_INTERNAL_MAX_SELF_END': 8,\n",
    "        'PRIMER_MAX_POLY_X': 100,\n",
    "        'PRIMER_INTERNAL_MAX_POLY_X': 100,\n",
    "        'PRIMER_SALT_MONOVALENT': 50.0,\n",
    "        'PRIMER_DNA_CONC': 50.0,\n",
    "        'PRIMER_MAX_NS_ACCEPTED': 0,\n",
    "        'PRIMER_THERMODYNAMIC_OLIGO_ALIGNMENT': 1,  # set 1 to actually use the thermodynamic calculations\n",
    "        'PRIMER_MAX_SELF_ANY': 12,\n",
    "        'PRIMER_MAX_SELF_END': 8,\n",
    "        'PRIMER_WT_SELF_END':1,   # use Primer_max_self_end\n",
    "        'PRIMER_MAX_SELF_END_TH': 30,\n",
    "        'PRIMER_WT_SELF_END_TH':1, # Primer_max_self_end_th\n",
    "        'PRIMER_PAIR_MAX_COMPL_ANY': 12,\n",
    "        'PRIMER_PAIR_MAX_COMPL_END': 8,\n",
    "        'PRIMER_PAIR_MAX_COMPL_ANY': 8,\n",
    "        'PRIMER_PAIR_WT_COMPL_ANY':2,\n",
    "        'PRIMER_PAIR_MAX_COMPL_ANY_TH': 30,\n",
    "        'PRIMER_PAIR_WT_COMPL_ANY_TH':2,\n",
    "        'PRIMER_MAX_HAIRPIN_TH': 47,\n",
    "        'PRIMER_WT_HAIRPIN_TH':1\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-18T08:46:52.330256Z",
     "start_time": "2019-09-18T08:46:41.468838Z"
    }
   },
   "outputs": [],
   "source": [
    "# for multicore get_primer_df has to be imported\n",
    "from p3_run import get_primer_df\n",
    "\n",
    "genome_split_folder = \"/Users/martinszyska/Dropbox/Icke/Work/static/genome/gatk/hg38/split\"\n",
    "\n",
    "test_primers = run_primer3(\n",
    "    test_df,\n",
    "    chroms_folder=genome_split_folder, # path to split chrome folder\n",
    "    pcr_config=PCR_config,\n",
    "    primer3_config=primer3_config,\n",
    "    threads=8         # multicore processing computes different chromosomes in parallel\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-11T07:19:48.251789Z",
     "start_time": "2019-09-11T07:19:48.147657Z"
    }
   },
   "outputs": [],
   "source": [
    "test_primers"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
